devito/ir/support/basic.py

```diff
@@ -621,6 +319,63 @@ def distance(self, other, findex=None):
         assert len(directions) == len(ret)
         return Vector(*[(-i) if d == Backward else i for i, d in zip(ret, directions)])
 
+    def touched_halo(self, findex):
+        """
+        Return a boolean 2-tuple, one entry for each ``findex`` DataSide. True
+        means that the halo is touched along that DataSide.
+        """
+        # If an irregularly (non-affine) accessed Dimension, conservatively
+        # assume the halo will be touched
+        if self.irregular(findex):
+            return (True, True)
+
+        d = self.aindices[findex]
+
+        # If the iterator is *not* a distributed Dimension, then surely the halo
+        # isn't touched
+        try:
+            if not d._maybe_distributed:
+                return (False, False)
+        except AttributeError:
+            assert is_integer(d)
+
+        # Given `d`'s iteration Interval `d[m, M]`, we know that `d` iterates between
+        # `d_m + m` and `d_M + M`
+        try:
+            m, M = self.intervals[d].offsets
+        except AttributeError:
+            if d.is_NonlinearDerived:
+                # We should only end up here with subsampled Dimensions
+                m, M = self.intervals[d.root].offsets
+            else:
+                # A constant (integer) is actually used to index into `findex`
+                assert is_integer(self[findex])
+                m, M = 0, 0
+
+        # If `m + (self[d] - d) < self.function._size_halo[d].left`, then `self`
```

Comment:
Maybe ( m+() ) ? Should you also use `findex` instead of `d`?

Reply:
No I think it's clear this way. The inner parentheses is just to emphasize the relationship between the access function (``self[d]``, e.g. "x + 3") and the dimension (e.g., ``x``)